Src/FluentAssertions/Common/MemberPathSegmentEqualityComparer.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace FluentAssertions.Common
+{
+    /// <summary>
+    /// Compares two segments of a <see cref="MemberPath"/>.
+    /// Sets the <see cref="AnyIndexQualifier"/> equal with any numeric index qualifier.
+    /// All other comparisons are default string equality.
+    /// </summary>
+    internal class MemberPathSegmentEqualityComparer : IEqualityComparer<string>
```

Comment:
🤔 When you check the latest `master` you'll see that we also started to allow `[]` as a valid separator for different purposes. I think it's worth combining that with this comparer.
🔧 I think it would be useful to have some examples in a `remarks` section. 
🔧 Maybe name the comparer to what it functionally does, e.g. something naming "indices"?

Reply:
❓Hm. If I use `[]` as valid `dottedPath` we can no longer use `SequenceEqual` to compare two `dottedPaths` because we are missing an element in the sequence. I. e. we get different path lengths for equivalent paths. I don't like that.

Example: 
```c#
"a[].b" // split: [a, b]
"a[1].b" // split: [a, 1, b]
```

Also with the introduction of `[]` the following is true and I belive this is a problem:
```c#
var path1 = new MemberPath("Parent.Property2.Test");
var path2 = new MemberPath("Parent.Property2");

path1.IsParentOrChildOf(path2).Should().BeTrue(); // correct

var path3 = new MemberPath("Parent[].Property2.Test");
var path4 = new MemberPath("Parent.Property2");

path3.IsParentOrChildOf(path4).Should().BeTrue(); // wrong: the paths are not parent / child
```

Reply:
Sorry for the late response, but needed to find some thing to think about the consequences. 

Is this a pure internal thing? In other words, this `*` will never leak into the consumer API? If so, then it's fine to keep it. If not, we need to change it.

Reply:
Hm, it is visible in the output when a test failed.
```
With configuration:
- Exclude member Level.Collection[*]Number
```

A consumer could do something like this
```c#
subject.Should().BeEquivalentTo(expected,
    options =>
    {
        options = options
            .Excluding(x => x.Level.Collection)
            .ThenExcluding(x => x.Number);

        var selectionRules = ((IEquivalencyAssertionOptions)options).SelectionRules;
        selectionRules.Last().ToString(); // --> "x.Level.Collection[*]Number"

        return options;
    });
```

Reply:
Then we need to change that, since it is inconsistent with the new `WithMapping` construct.

Reply:
Okay so I removed the `*` from the `dottedPath`. I add it back later when splitting the `dottedPath` into segments. The segments are private to the `MemberPath` and also not returned by any function.

A failed assertion returns something like this:
```
With configuration:
- Exclude member Collection[]Number
```